packages/SwingSet/src/liveslots/virtualReferences.js

```diff
@@ -466,43 +466,56 @@ export function makeVirtualReferenceManager(
   /** @type {Map<string, Set<Recognizer>>} */
   const vrefRecognizers = new Map();
 
-  function addRecognizableValue(value, recognizer) {
+  function addRecognizableValue(value, recognizer, recognizerIsVirtual) {
     const vref = getSlotForVal(value);
     if (vref) {
       const { type, allocatedByVat, virtual } = parseVatSlot(vref);
       if (type === 'object' && (!allocatedByVat || virtual)) {
-        let recognizerSet = vrefRecognizers.get(vref);
-        if (!recognizerSet) {
-          recognizerSet = new Set();
-          vrefRecognizers.set(vref, recognizerSet);
+        if (recognizerIsVirtual) {
+          syscall.vatstoreSet(`vom.ir.${vref}|${recognizer}`, '1');
+        } else {
+          let recognizerSet = vrefRecognizers.get(vref);
+          if (!recognizerSet) {
+            recognizerSet = new Set();
+            vrefRecognizers.set(vref, recognizerSet);
+          }
+          recognizerSet.add(recognizer);
         }
-        recognizerSet.add(recognizer);
       }
     }
   }
 
-  function removeRecognizableVref(vref, recognizer) {
+  function removeRecognizableVref(vref, recognizer, recognizerIsVirtual) {
     const { type, allocatedByVat, virtual } = parseVatSlot(vref);
     if (type === 'object' && (!allocatedByVat || virtual)) {
-      const recognizerSet = vrefRecognizers.get(vref);
-      assert(recognizerSet && recognizerSet.has(recognizer));
-      recognizerSet.delete(recognizer);
-      if (recognizerSet.size === 0) {
-        vrefRecognizers.delete(vref);
-        if (!allocatedByVat) {
-          addToPossiblyRetiredSet(vref);
+      if (recognizerIsVirtual) {
+        syscall.vatstoreDelete(`vom.ir.${vref}|${recognizer}`);
+      } else {
+        const recognizerSet = vrefRecognizers.get(vref);
+        assert(recognizerSet && recognizerSet.has(recognizer));
+        recognizerSet.delete(recognizer);
+        if (recognizerSet.size === 0) {
+          vrefRecognizers.delete(vref);
+          if (!allocatedByVat) {
+            addToPossiblyRetiredSet(vref);
+          }
         }
       }
     }
   }
 
-  function removeRecognizableValue(value, recognizer) {
+  function removeRecognizableValue(value, recognizer, recognizerIsVirtual) {
     const vref = getSlotForVal(value);
     if (vref) {
-      removeRecognizableVref(vref, recognizer);
+      removeRecognizableVref(vref, recognizer, recognizerIsVirtual);
     }
   }
 
+  let deleteCollectionEntry;
```

Comment:
This bit of circularity is unfortunate, but I imagine this is the best available way to deal with it.

Reply:
Yes. I muttered under my breath about this at length when I was unable to think of a better way.